plugins/Tools/RotateTool/RotateTool.py

```diff
@@ -151,3 +156,99 @@ def setRotationSnapAngle(self, angle):
+
+    def layFlat(self):
+        # Based on https://github.com/daid/Cura/blob/SteamEngine/Cura/util/printableObject.py#L207
+        # Note: Y & Z axis are swapped
+
+        self.operationStarted.emit(self)
+        progress_message = Message("Laying object flat on buildplate...", lifetime = 0, dismissable = False)
+        progress_message.setProgress(0)
+        progress_message.show()
+
+        selected_object = Selection.getSelectedObject(0)
+        transformed_vertices = selected_object.getMeshDataTransformed().getVertices()
+        min_y_vertex = transformed_vertices[transformed_vertices.argmin(0)[1]]
+        dot_min = 1.0
+        dot_v = None
+
+        iterations = 0
+        total_iterations = len(transformed_vertices) * 2
+        last_progress = 0
+
+        for v in transformed_vertices:
+            diff = v - min_y_vertex
+            length = math.sqrt(diff[0] * diff[0] + diff[2] * diff[2] + diff[1] * diff[1])
+            if length < 5:
+                continue
+            dot = (diff[1] / length)
+            if dot_min > dot:
+                dot_min = dot
+                dot_v = diff
+            iterations += 1
+            progress = round(100*iterations/total_iterations)
+            if progress != last_progress:
+                last_progress = progress
+                progress_message.setProgress(progress)
+                QApplication.processEvents()
+
+        if dot_v is None:
+            progress_message.hide()
+            self.operationStopped.emit(self)
+            return
+
+        rad = -math.asin(dot_min)
+        m = Matrix([
+            [ math.cos(rad), math.sin(rad), 0 ],
+            [-math.sin(rad), math.cos(rad), 0 ],
+            [ 0,             0,             1 ]
+        ])
+        selected_object.rotate(Quaternion.fromMatrix(m), SceneNode.TransformSpace.Parent)
+
+        rad = math.atan2(dot_v[2], dot_v[0])
+        m = Matrix([
+            [ math.cos(rad), 0, math.sin(rad)],
+            [ 0,             1, 0 ],
+            [-math.sin(rad), 0, math.cos(rad)]
+        ])
+        selected_object.rotate(Quaternion.fromMatrix(m), SceneNode.TransformSpace.Parent)
+
+        transformed_vertices = selected_object.getMeshDataTransformed().getVertices()
+        min_y_vertex = transformed_vertices[transformed_vertices.argmin(0)[1]]
+        dot_min = 1.0
+        dot_v = None
+
+        for v in transformed_vertices:
+            diff = v - min_y_vertex
+            length = math.sqrt(diff[2] * diff[2] + diff[1] * diff[1])
+            if length < 5:
+                continue
+            dot = (diff[1] / length)
+            if dot_min > dot:
+                dot_min = dot
+                dot_v = diff
+            iterations += 1
+            progress = round(100*iterations/total_iterations)
+            if progress != last_progress:
+                last_progress = progress
+                progress_message.setProgress(progress)
+                QApplication.processEvents()
```

Comment:
I am not too happy with the introduced dependency on QApplication here, though I suppose for the moment we do not really have a different solution since we lack support for "modal jobs" that block interaction with the build plate.